### ISSUE TITLE
Distributed training of slim examples use only one node

### ISSUE THREAD

**Issue by classicsong** [tensorflow/models#3945]
I deployed the slim example on my distributed GPU environment for distributed training (with 2 nodes, each has 4 P40 GPU). When using async training mode, I found out that only one node has GPU workload an other node's GPU is idle.

TensorFlow Version: 1.6.0
TensorFlow Model Version: 1.8.0
cuda version: nvidia-docker with cuda9.1 and cudnn7
os: ubuntu 16.04
python 2.7.0

Distributed Env config:
ps: "A:2222", "B:2222"
worker: "A:2223", "B:2223"
PS use CPU only and worker use GPU with num_clones=4

The solution is to modify the clone_device(self, clone_index) function in deployment/model_deploy.py
to assign task id with device assignment.

    --- a/research/slim/deployment/model_deploy.py
    +++ b/research/slim/deployment/model_deploy.py
    @@ -593,6 +593,9 @@ class DeploymentConfig(object):
           device += '/device:CPU:0'
         else:
           device += '/device:GPU:%d' % clone_index
    +
    +    if self._num_ps_tasks > 0:
    +      device = '%s/task:%d' % (device, self._replica_id) 
         return device
 
       def clone_scope(self, clone_index):

**Comment by Lynkzhang** [comment]
hi, can you tell me where to set the Distributed Env config.  I have no idea how to run slim code on multiple node.. thank you very much!

**Comment by classicsong** [comment]
> hi, can you tell me where to set the Distributed Env config. I have no idea how to run slim code on multiple node.. thank you very much!

https://www.tensorflow.org/deploy/distributed Here is the official description.
If you can read chinese, here is the dist intro of our platform https://docs.ucloud.cn/ai/uai-train/dist/tensorflow. It includes how to organize the TF_CONFIG

**Comment by Lynkzhang** [comment]
Thank you very much for your kind reply! I have read these documents before. I know I have to give tf.train.ClusterSpec, the problem for me is that , I have no idea how to modify the example training code provided by slim as well as how to excute it.

**Comment by FrankTsui** [comment]
@Lynkzhang , Hi, have you solved the problem that how to set distributed config parameters (slim/deployment/model_deploy.py) to realize distributed training with tf? Thank you very much.

**Comment by Lynkzhang** [comment]
Basically I didn't use the APIs, you have to modify trainning code, follow example showed in TF website... Somehow the API for distributed optimizor have bug and already decrepited ... so if maybe you need horovod. As @classicsong mentioned.  But I have to say, now I give up slim.. too old.

**Comment by FrankTsui** [comment]
@Lynkzhang , thank you. But now, how to implement distributed training with multiple replicas and multiple clones?

**Comment by Lynkzhang** [comment]
What I mean is , there exists bug, so ... AFAK, no solution

**Comment by FrankTsui** [comment]
@Lynkzhang , yeah, thanks, but don't you use distributed training now?

**Comment by Lynkzhang** [comment]
@FrankTsui 　I used horovod

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
